contribs/gmf/src/services/snapping.js

```diff
@@ -0,0 +1,583 @@
+    this.unregisterAllTreeCtrl_();
+    keys.forEach(ol.events.unlistenByKey);
+    keys.length = 0;
+  }
+
+  this.map_ = map;
+
+  if (map) {
+    this.treeCtrlsUnregister_ = this.rootScope_.$watchCollection(function() {
+      if (this.gmfTreeManager_.rootCtrl) {
+        return this.gmfTreeManager_.rootCtrl.children;
+      }
+    }.bind(this), function(value) {
+      // Timeout required, because the collection event is fired before the
+      // leaf nodes are created and they are the ones we're looking for here.
+      this.timeout_(function() {
+        if (value) {
+          this.unregisterAllTreeCtrl_();
+          this.gmfTreeManager_.rootCtrl.traverseDepthFirst(this.registerTreeCtrl_.bind(this));
+        }
+      }.bind(this), 0);
+    }.bind(this));
+
+    keys.push(
+      ol.events.listen(
+        this.gmfThemes_,
+        gmf.ThemesEventType.CHANGE,
+        this.handleThemesChange_,
+        this
+      )
+    );
+
+    var view = map.getView();
+    keys.push(
+      ol.events.listen(
+        view,
+        ol.Object.getChangeEventType(ol.View.Property.CENTER),
+        this.handleMapViewChange_,
+        this
+      )
+    );
+    keys.push(
+      ol.events.listen(
+        view,
+        ol.Object.getChangeEventType(ol.View.Property.RESOLUTION),
+        this.handleMapViewChange_,
+        this
+      )
+    );
+  }
+};
+
+
+/**
+ * Called when the themes change. Get the OGC servers, then listen to the
+ * tree manager Layertree controllers array changes.
+ * @private
+ */
+gmf.Snapping.prototype.handleThemesChange_ = function() {
+  this.ogcServers_ = null;
+  this.gmfThemes_.getOgcServersObject().then(function(ogcServers) {
+    this.ogcServers_ = ogcServers;
+  }.bind(this));
+};
+
+
+/**
+ * Registers a newly added Layertree controller 'leaf'. If it's snappable,
+ * create and add a cache item with every configuration required to do the
+ * snapping. It becomes active when its state is set to 'on'.
+ *
+ * @param {ngeo.LayertreeController} treeCtrl Layertree controller to register
+ * @private
+ */
+gmf.Snapping.prototype.registerTreeCtrl_ = function(treeCtrl) {
+
+  // Skip any Layertree controller that has a node that is not a leaf
+  var node = /** @type {GmfThemesGroup|GmfThemesLeaf} */ (treeCtrl.node);
+  if (node.children) {
+    return;
+  }
+
+  // If treeCtrl is snappable and supports WFS, listen to its state change.
+  // When it becomes visible, it's added to the list of snappable tree ctrls.
+  var snappingConfig = gmf.LayertreeController.getSnappingConfig(treeCtrl);
+  if (snappingConfig) {
+    var wfsConfig = this.getWFSConfig_(treeCtrl);
+    if (wfsConfig) {
+      var uid = goog.getUid(treeCtrl);
+
+      var stateWatcherUnregister = this.rootScope_.$watch(
+        function() {
+          return treeCtrl.getState();
+        }.bind(this),
+        this.handleTreeCtrlStateChange_.bind(this, treeCtrl)
+      );
+
+      // Todo: some of the properties here are hardcoded, but could come from
```

Comment:
If we decide to merge this pull-request, we must open an issue to not forget this todo